jobsdb/unionQuery.go

```diff
@@ -0,0 +1,284 @@
+}
+
+//Unprocessed
+
+func (jd *HandleT) getUnprocessedUnionQuerystring(unionMap map[string]int, ds dataSetT, params GetQueryParamsT) string {
+	var queries []string
+
+	for customer, count := range unionMap {
+		//do cache stuff here
+		queries = append(queries, jd.getSingleCustomerUnprocessedQueryString(customer, count, ds, params, true))
+	}
+
+	return `(` + strings.Join(queries, `) UNION (`) + `)`
+}
+
+func (jd *HandleT) getSingleCustomerUnprocessedQueryString(customer string, count int, ds dataSetT, params GetQueryParamsT, order bool) string {
+	customValFilters := params.CustomValFilters
+	parameterFilters := params.ParameterFilters
+	var sqlStatement string
+
+	if useJoinForUnprocessed {
+		// event_count default 1, number of items in payload
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				`FROM %[1]s AS jobs `+
+				`LEFT JOIN %[2]s AS job_status ON jobs.job_id=job_status.job_id `+
+				`WHERE job_status.job_id is NULL AND customer='%[3]s'`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	} else {
+		sqlStatement = fmt.Sprintf(
+			selectQuery+
+				` FROM %[1]s AS jobs `+
+				`WHERE customer='%[3]s' AND jobs.job_id NOT IN (SELECT DISTINCT(job_status.job_id) FROM %[2]s AS job_status)`,
+			ds.JobTable, ds.JobStatusTable, customer)
+	}
+
+	if len(customValFilters) > 0 && !params.IgnoreCustomValFiltersInQuery {
+		sqlStatement += " AND " + constructQuery(jd, "jobs.custom_val", customValFilters, "OR")
+	}
+
+	if len(parameterFilters) > 0 {
+		sqlStatement += " AND " + constructParameterJSONQuery("jobs", parameterFilters)
+	}
+
+	//avoinding AfterJobID for now
+
+	if params.UseTimeFilter {
+		sqlStatement += fmt.Sprintf(" AND created_at < %s", params.Before)
+	}
+
+	if order {
+		sqlStatement += " ORDER BY jobs.job_id"
+	}
+
+	if count > 0 {
+		sqlStatement += fmt.Sprintf(" LIMIT %d", count)
+	}
+
+	if params.EventCount > 0 {
+		sqlStatement = fmt.Sprintf(`SELECT * FROM (`+sqlStatement+`) AS subquery WHERE running_event_counts - event_count + 1 <= %d;`, params.EventCount)
+	}
+
+	return sqlStatement
+}
+
+func (jd *HandleT) GetUnprocessedUnion(unionMap map[string]int, params GetQueryParamsT) []*JobT {
+	// if params.JobCount == 0 { //how do we use this check..? is it required anymore
+	// 	return []*JobT{}
+	// }
+
+	//add stats
+
+	//The order of lock is very important. The migrateDSLoop
+	//takes lock in this order so reversing this will cause
+	//deadlocks
+	jd.dsMigrationLock.RLock()
+	jd.dsListLock.RLock()
+	defer jd.dsMigrationLock.RUnlock()
+	defer jd.dsListLock.RUnlock()
+
+	dsList := jd.getDSList(false)
+	outJobs := make([]*JobT, 0)
+
+	//removed count assert, because that params.count is not used..?
+
+	for _, ds := range dsList {
+		jobs := jd.getUnprocessedUnionDS(ds, unionMap, params)
+		outJobs = append(outJobs, jobs...)
+		// count -= len(jobs) //this is not necessary anymore right?
+
+		//we could further limit jobs by eventCount - ignoring that for now
+		// just focusing on getting so many rows of jobs for now
```

Comment:
We never store multiple events per job in the `router` / `batchrouter` tables. We want to have one jobs per event so we can handle failures and retries independently.

Thus, event count is not necessary here. Plus it is going to introduce a lot of complexity. My opinion is to not implement this.

Reply:
oh yes, that missed my mind. Won't go ahead with that anymore.